clients/src/main/java/org/apache/kafka/common/MetricNameTemplate.java

```diff
@@ -26,47 +26,86 @@
  * A template for a MetricName. It contains a name, group, and description, as
  * well as all the tags that will be used to create the mBean name. Tag values
  * are omitted from the template, but are filled in at runtime with their
- * specified values.
+ * specified values. The order of the tags is maintained, if an ordered set
+ * is provided, so that the mBean names can be compared and sorted lexicographically.
  */
 public class MetricNameTemplate {
     private final String name;
     private final String group;
     private final String description;
-    private Set<String> tags;
+    private LinkedHashSet<String> tags;
 
-    public MetricNameTemplate(String name, String group, String description, Set<String> tags) {
+    /**
+     * Create a new template. Note that the order of the tags will be preserved if the supplied
+     * {@code tagsNames} set has an order.
+     *
+     * @param name the name of the metric; may not be null
+     * @param group the name of the group; may not be null
+     * @param description the description of the metric; may not be null
+     * @param tagsNames the set of metric tag names, which can/should be a set that maintains order; may not be null
```

Comment:
Interesting. If someone passes in a non-order-preserving Set, then does that mean that the order of the MBeans in the docs might change from one build to the next? The build of the docs won't be repeatable, then?

Reply:
Technically, it is possible that the order is different on different JVMs.

Reply:
BTW, this is true even without the PR.

Reply:
Kinda, but it had a TreeMap before, so it used the "natural ordering" of the Strings which means... something.

Anyway, are we okay with the fact that the ordering may change according to the person who built the docs? The kafka.apache.org webpage will then show whatever order the site-builder's JVM used? Or is it Jenkins?

We don't really have any other choice, other than making everyone use LinkedHashSets, right? And we decided we don't want that.

Reply:
True that before it used `TreeMap`. I guess I'd like to get some other feedback, perhaps from @ijuma and @rajinisivaram about whether it'd be better to make all the changes to enforce order.

Reply:
Unfortunately this is public API, so just changing it across the board may not be trivial and requires a KIP.

I'd suggest getting the quick documentation win via this PR (modulo my other comment) and deal with any more extensive change in a follow up JIRA.